### DISCUSSION TITLE
Questions around SSR

### DISCUSSION THREAD

**Discussion by zhuangdaz** [urql-graphql/urql#2088] [Q&A] [answered]
I find the document about SSR could be more clear.

For example, the following paragraph in the [NextJS](https://formidable.com/open-source/urql/docs/advanced/server-side-rendering/#nextjs) section is confusing because I don't see  `neverSuspend` is set to true in any of your provided [examples](https://github.com/FormidableLabs/urql/tree/main/examples/with-next).
>When you are using getStaticProps, getServerSideProps, or getStaticPaths, you should opt-out of Suspense by setting the neverSuspend option to true in your withUrqlClient configuration. During the prepass of your component tree next-urql can't know how these functions will alter the props passed to your page component. This injection could change the variables used in your useQuery. This will lead to error being thrown during the subsequent toString pass, which isn't supported in React 16.

Another confusion I have is `ssr` option.
In this paragraph, I am not sure what is enabled by setting `ssr` to true. Could you elaborate on that?
>Unless the component that is being wrapped already has a getInitialProps method, next-urql won't add its own SSR logic, which automatically fetches queries during server-side rendering. This can be explicitly enabled by passing the { ssr: true } option as a second argument to withUrqlClient.

And also it sounds contradictory to the first statement in the [Using getStaticProps or getServerSideProps](https://formidable.com/open-source/urql/docs/advanced/server-side-rendering/#using-getstaticprops-or-getserversideprops) section:
>By default withUrqlClient will add getInitialProps to the component you're wrapping it in

Finally, I wonder if we can rename `ssr` field to `disableGetInitialProps` since  `getStaticProps` and `getServerSideProps` are also SSR and setting `ssr: false` for their usages doesn't make sense.

**Comment by JoviDeCroock** [comment, marked as answer]
Hey,

It's called `ssr` because when you toggle it to `true` you are telling `next-urql` that you want it to handle ssr for you, that's why we use `getInitialProps` when it's turned on. We'll wrap your component with `getInitialProps` and run a [prepass](https://github.com/FormidableLabs/react-ssr-prepass) which will get all the data for the tree you are rendering and send it over to the client.

When we have this on, we'll leverage `suspense` so we throw Promises upwards so it can actually resolve them and handle everything deterministically on the server. `neverSuspend` is an escape hatch for when you are using a query that isn't part of the server prepass but is of the stringification.

When you turn `ssr` off, you're telling the client that you are facilitating what happens on the server and that the client only has to care about rehydrating that data when it's on the client. This is an important distinction and reasoning behind the naming scheme.

**Reply by zhuangdaz** [reply]
That makes much more sense to me now. Ty!

I have a page that shows a login'ed user's active meetings(using `getServerSideProps`) as well as a form to post a new meeting.
Since I need a `authExchange` for authentication, I have to manually set `exchanges`. Should I reuse the `ssr` from server or create a new one? 
```
export default withUrqlClient((ssr) => ({
  url: config.api.aws_appsync_graphqlEndpoint,
  exchanges: [
    devtoolsExchange,
    dedupExchange,
    cacheExchange,
    ssr, // should we reuse ssr from server or create a new one?
    authExchange({
      getAuth,
      addAuthToOperation,
    }),
    fetchExchange,
  ],
}), { ssr: false })(Meetings)
```

**Reply by JoviDeCroock** [reply]
The reuse here is advised because this will be the populated ssr-exchange for the client-side. On the client we'll try to hydrate your cache [like this](https://github.com/FormidableLabs/urql/blob/main/packages/next-urql/src/with-urql-client.ts#L65) which will contain the base-data from your `getServerSideProps` call because you've forwarded this into our special place in [`pageProps`](https://github.com/FormidableLabs/urql/blob/main/packages/next-urql/src/with-urql-client.ts#L56)

**Reply by zhuangdaz** [reply]
I see. So behind the scene, you will construct a new ssr based on the populated one and also having `isClient` set to `true`.

**Comment by zhuangdaz** [comment]
I want to correct my earlier misunderstanding:
>For example, the following paragraph in the NextJS section is confusing because I don't see neverSuspend is set to true in any of your provided examples.

This is not true since the second parameter for `initUrqlClient` is `canEnableSuspense` and it is set to `false`.
